timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/client/DistributingModulusGeneratorTest.java

```diff
@@ -0,0 +1,139 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.transaction.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.junit.Test;
+
+import com.google.common.collect.Iterables;
+
+public class DistributingModulusGeneratorTest {
+    private DistributingModulusGenerator generator;
+
+    @Test
+    public void doesNotReuseResiduesIfNotNeeded() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+    }
+
+    @Test
+    public void waitsToReuseResiduesIfNeeded() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+    }
+
+    @Test
+    public void throwsIfUnmarkingUnmarkedResidue() {
+        givenGeneratorHasModulus(3);
+
+        assertThatThrownBy(() -> unmarkResidue(0)).satisfies(this::failureArisesFromUnmarkingResidues);
+    }
+
+    @Test
+    public void canUnmarkMarkedResidues() {
+        givenGeneratorHasModulus(3);
+
+        requestResidues(3);
+        unmarkResidue(0);
+        unmarkResidue(1);
+        unmarkResidue(2);
+        // Not throwing is a success!
+    }
+
+    @Test
+    public void cannotDoubleUnmarkMarkedResidues() {
+        givenGeneratorHasModulus(3);
+
+        requestResidues(3);
+
+        unmarkResidue(0);
+        assertThatThrownBy(() -> unmarkResidue(0)).satisfies(this::failureArisesFromUnmarkingResidues);
+    }
+
+    @Test
+    public void reassignsUnmarkedResiduesIfTheyAreNowLeastReferenced() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        unmarkResidue(1);
+        assertThat(requestResidues(1)).containsExactly(1);
+    }
+
+    @Test
+    public void repeatedlyReassignsUnmarkedResidues() {
+        givenGeneratorHasModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+
+        unmarkResidue(2);
+        unmarkResidue(2);
+        unmarkResidue(2);
+        List<Integer> additionalResponses = requestResidues(6);
+
+        assertThat(additionalResponses.subList(0, 3)).containsExactly(2, 2, 2);
+        assertListBalanced(additionalResponses.subList(3, 6), 3);
+    }
+
+    private void givenGeneratorHasModulus(int modulus) {
+        generator = new DistributingModulusGenerator(modulus);
+    }
+
+    private List<Integer> requestResidues(int times) {
+        return IntStream.range(0, times)
+                .map(unused -> generator.getAndMarkResidue())
+                .boxed()
+                .collect(Collectors.toList());
+    }
+
+    private void requestResiduesAndAssertBalanced(int times) {
```

Comment:
If you hardcode generator, then this should also be hardcoded for 3

Reply:
(Didn't hardcode the generator.)

Reply:
is it merely coincidence that `times` and `modulus` are equal?

Reply:
The verifying method only works if times is a multiple of modulus